src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -188,47 +190,349 @@ bool SotaUptaneClient::updateMeta() {
+
+    int local_version;
+    std::string images_snapshot_stored;
+    if (storage->loadRole(&images_snapshot_stored, Uptane::RepositoryType::Images, Uptane::Role::Snapshot())) {
+      local_version = Uptane::extractVersionUntrusted(images_snapshot_stored);
+    } else {
+      local_version = -1;
+    }
+
+    if (local_version > remote_version) {
+      return false;
+    } else if (local_version == remote_version) {
+      return true;
+    }
+
+    if (!images_repo.verifySnapshot(images_snapshot)) {
+      last_exception = images_repo.getLastException();
+      return false;
+    }
+
+    storage->storeRole(images_snapshot, Uptane::RepositoryType::Images, Uptane::Role::Snapshot(),
+                       Uptane::Version(remote_version));
+
+    if (images_repo.snapshotExpired()) {
+      last_exception = Uptane::ExpiredMetadata("repo", "snapshot");
+      return false;
+    }
+  }
+
+  // Update Images Targets Metadata
+  {
+    std::string images_targets;
+
+    if (!uptane_fetcher.fetchRole(&images_targets, Uptane::RepositoryType::Images, Uptane::Role::Targets())) {
+      return false;
+    }
+    int remote_version = Uptane::extractVersionUntrusted(images_targets);
+
+    int local_version;
+    std::string images_targets_stored;
+    if (storage->loadRole(&images_targets_stored, Uptane::RepositoryType::Images, Uptane::Role::Targets())) {
+      local_version = Uptane::extractVersionUntrusted(images_targets_stored);
+    } else {
+      local_version = -1;
+    }
+
+    if (local_version > remote_version) {
+      return false;
+    } else if (local_version == remote_version) {
+      return true;
+    }
+    if (!images_repo.verifyTargets(images_targets)) {
+      last_exception = images_repo.getLastException();
+      return false;
+    }
+    storage->storeRole(images_targets, Uptane::RepositoryType::Images, Uptane::Role::Targets(),
+                       Uptane::Version(remote_version));
+
+    if (images_repo.targetsExpired()) {
+      last_exception = Uptane::ExpiredMetadata("repo", "targets");
+      return false;
+    }
+  }
+  return true;
+}
+
+bool SotaUptaneClient::getNewTargets(std::vector<Uptane::Target> *new_targets) {
+  std::vector<Uptane::Target> targets = director_repo.getTargets();
+  for (auto targets_it = targets.cbegin(); targets_it != targets.cend(); ++targets_it) {
+    bool is_new = true;
+    for (auto ecus_it = targets_it->ecus().cbegin(); ecus_it != targets_it->ecus().cend(); ++ecus_it) {
+      Uptane::EcuSerial ecu_serial = ecus_it->first;
+      Uptane::HardwareIdentifier hw_id = ecus_it->second;
+      auto hwid_it = hw_ids.find(ecu_serial);
+      if (hwid_it == hw_ids.end()) {
+        LOG_WARNING << "Unknown ECU ID in director targets metadata: " << ecu_serial.ToString();
+        is_new = false;
+        break;
+      }
+
+      if (hwid_it->second != hw_id) {
+        LOG_ERROR << "Wrong hardware identifier for ECU " << ecu_serial.ToString();
+        return false;
```

Comment:
minor, but `new_targets` still get modified if a target has an update and this condition is triggered in another loop iteration. This won't cause a problem in a regular use of this method (`new_targets` is most likely a fresh object) but it's a subtle breach of the function's contract.

We could start using `boost::optional<T>` as a return type for this kind of methods, when possible.

Reply:
Yes, it's true, I assumed that we can't trust contents of `new_targets` if the return value is false.
You're right about `boost::optional`, it seems to be around for quite a long time, so we shouldn't have compiler compatibility issues. As much as I'd prefer to just use the latest C++ standard.

Reply:
Let's do it in a separate PR, I'll want to change some interfaces in INvStorage in this way as well.